components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp

```diff
@@ -303,10 +144,1035 @@ void mixing_ratio_to_cloud_mass(
+#endif // RRTMGP_ENABLE_YAKL
+
+// New interface for Kokkos and flexible types
 #ifdef RRTMGP_ENABLE_KOKKOS
+template <typename RealT=Real, typename LayoutT=Kokkos::LayoutRight, typename DeviceT=DefaultDevice>
+struct rrtmgp_interface {
+
+using MDRP = typename conv::MDRP<LayoutT>;
+
+template <typename T>
+using view_t = Kokkos::View<T, LayoutT, DeviceT>;
+
+template <typename T>
+using oview_t = Kokkos::Experimental::OffsetView<T, LayoutT, DeviceT>;
+
+template <typename T>
+using hview_t = Kokkos::View<T, LayoutT, HostDevice>;
+
+using pool_t = conv::MemPoolSingleton<RealT, DeviceT>;
+
+/*
+ * Objects containing k-distribution information need to be initialized
+ * once and then persist throughout the life of the program, so we
+ * declare them here within the rrtmgp namespace.
+ */
+static inline GasOpticsRRTMGPK<RealT, LayoutT, DeviceT> k_dist_sw_k;
+static inline GasOpticsRRTMGPK<RealT, LayoutT, DeviceT> k_dist_lw_k;
+
+/*
+ * Objects containing cloud optical property look-up table information.
+ * We want to initialize these once and use throughout the life of the
+ * program, so declare here and read data in during rrtmgp_initialize().
+ */
+static inline CloudOpticsK<RealT, LayoutT, DeviceT> cloud_optics_sw_k;
+static inline CloudOpticsK<RealT, LayoutT, DeviceT> cloud_optics_lw_k;
+
+/*
+ * Flag to indicate whether or not we have initialized RRTMGP
+ */
+static inline bool initialized_k = false;
+
+/*
+ * Initialize data for RRTMGP driver
+ */
+static void rrtmgp_initialize(
+  GasConcsK<RealT, LayoutT, DeviceT> &gas_concs,
+  const std::string& coefficients_file_sw, const std::string& coefficients_file_lw,
+  const std::string& cloud_optics_file_sw, const std::string& cloud_optics_file_lw,
+  const std::shared_ptr<spdlog::logger>& logger)
+{
+  // If we've already initialized, just exit
+  if (initialized_k) {
+    if (logger)
+      logger->info("RRTMGP is already initialized; skipping\n");
+    return;
+  }
+
+  // Initialize Kokkos
+  if (!Kokkos::is_initialized()) {  Kokkos::initialize(); }
+
+  // Load and initialize absorption coefficient data
+  load_and_init(k_dist_sw_k, coefficients_file_sw, gas_concs);
+  load_and_init(k_dist_lw_k, coefficients_file_lw, gas_concs);
+
+  // Load and initialize cloud optical property look-up table information
+  load_cld_lutcoeff(cloud_optics_sw_k, cloud_optics_file_sw);
+  load_cld_lutcoeff(cloud_optics_lw_k, cloud_optics_file_lw);
+
+  // initialize kokkos rrtmgp pool allocator
+  const size_t base_ref = 18000;
+  const size_t ncol = gas_concs.ncol;
+  const size_t nlay = gas_concs.nlay;
+  const size_t nlev = SCREAM_NUM_VERTICAL_LEV;
+  const size_t my_size_ref = ncol * nlay * nlev;
+  pool_t::init(2e6 * (float(my_size_ref) / base_ref));
+
+  // We are now initialized!
+  initialized_k = true;
+}
+
+/*
+ * Compute band-by-band surface albedos from broadband albedos.
+ */
+template <typename SfcAlbDirVisT, typename SfcAlbDirNirT, typename SfcAlbDifVisT,
```

Comment:
Since we already know the device and layout, it's possible these views didn't have to be templates.

Reply:
Aren't these all the same type? Why template on each individual type if in use they are all the same? I.e., why couldn't we do:
```
template<typename T, typename S> static void compute_band_by_band_surface_albedos(
  const int incol, const int nswbands,
  T & sfc_alb_dir_vis, T &sfc_alb_dir_nir, T &sfc_alb_dif_vis, T &sfc_alb_dif_nir,
  S &sfc_alb_dir, S &sfc_alb_dif)
```